modules/asset-manager-impl/src/main/java/org/opencastproject/assetmanager/impl/AbstractAssetManagerWithTieredStorage.java

```diff
@@ -0,0 +1,412 @@
+    return Enrichments.enrich(query.run());
+  }
+
+  public void moveSnapshotsByIdAndDate(final String mpId, final Date start, final Date end, final String targetStore)
+          throws NotFoundException {
+    RichAResult results = getSnapshotsByDate(start, end);
+
+    if (results.getRecords().isEmpty()) {
+      throw new NotFoundException("No media package with id " + mpId + " found between " + start + " and " + end);
+    }
+
+    processOperations(results, targetStore);
+  }
+
+  /** Returns true if the store id is equal to the local asset store's id */
+  private boolean isLocalAssetStoreId(String storeId) {
+    return getLocalAssetStore().getStoreType().equals(storeId);
+  }
+
+  /** Returns true if the store id is not equal to the local asset store's id */
+  private boolean isRemoteAssetStoreId(String storeId) {
+    return !isLocalAssetStoreId(storeId);
+  }
+
+  @Override
+  public Opt<Asset> getAsset(Version version, final String mpId, final String mpeId) {
+    // try to fetch the asset
+    for (final AssetDtos.Medium asset : getDb().getAsset(RuntimeTypes.convert(version), mpId, mpeId)) {
+      for (final String storageId : getSnapshotStorageLocation(version, mpId)) {
+        for (final AssetStore store : getAssetStore(storageId)) {
+          for (final InputStream assetStream : store.get(StoragePath.mk(asset.getOrganizationId(), mpId, version, mpeId))) {
+            final Asset a = new AssetImpl(
+                    AssetId.mk(version, mpId, mpeId),
+                    assetStream,
+                    asset.getAssetDto().getMimeType(),
+                    asset.getAssetDto().getSize(),
+                    asset.getStorageId(),
+                    asset.getAvailability());
+            return Opt.some(a);
+          }
+        }
+      }
+    }
+    return Opt.none();
+  }
+
+  /** Move the assets for a snapshot to the target store */
+  private void copyAssetsToStore(Snapshot snap, AssetStore store) throws Exception {
+    final String mpId = snap.getMediaPackage().getIdentifier().toString();
+    final String orgId = snap.getOrganizationId();
+    final Version version = snap.getVersion();
+    final String prettyMpId = mpId + "@v" + version;
+    logger.debug(format("Moving assets for snapshot %s to store %s", prettyMpId, store.getStoreType()));
+    for (final MediaPackageElement e : snap.getMediaPackage().getElements()) {
+      if (!MOVABLE_TYPES.contains(e.getElementType())) {
+        logger.debug(format("Skipping %s because type is %s", e.getIdentifier(), e.getElementType()));
+        continue;
+      }
+      logger.debug(format("Moving %s to store %s", e.getIdentifier(), store.getStoreType()));
+      final StoragePath storagePath = StoragePath.mk(orgId, mpId, version, e.getIdentifier());
+      final Opt<StoragePath> existingAssetOpt = findAssetInVersionsAndStores(e.getChecksum().toString(), store.getStoreType());
```

Comment:
In my tests, findAssetInVersionsAndStores does not find the asset previously archived to the aws-s3 store when there's another version of the asset that is "local" because the "local" asset is always returned. I tracked it to the findAssetByChecksum that uses the "singleResult" clause so the first one found is returned.
This causes several copies of the asset in aws-s3 instead of only one.